torch/csrc/distributed/c10d/NanCheck.cu

```diff
@@ -79,32 +79,66 @@ struct CheckBytePack<T, /*EltPerPack*/8> {
   }
 };
 
-// (v) Template specialization for Float8_e4m3fn.
+// (v) Template specialization for Float8 types.
 // EltPerPack = 16 / 1 = 16
 
+// We want to check 8 x FP8 simultaneously, hence this template definition.
+template<typename T>
+struct HasNanFP8x8 {
+  // I am a dumb implementation. You should never call in here, unless the check
+  // for your datatype hasn't been implemented. File a Feature Request in that
+  // case.
+  static __device__ __forceinline__ bool check(uint64_t fp8x8) {
+    T* data = &fp8x8;
+    bool hasNan = false;
+    for (int i = 0; i < 8; i++) {
+      hasNan |= isnan(data[i]);
+    }
+    return hasNan;
+  }
+};
+
 // isnan condition for Float8_e4m3fn:
 // (x & 0b01111111) == 0b01111111
 // i.e.
 // (x & 0x7f) == 0x7f
 
-// We want to check 8 x FP8 simultaneously. The algorithm is as follows:
+// The algorithm is as follows:
 // (1) Mask out the most significant bit with mask 0x7f.
 // (2) If the result is 0x7f (is nan), the following arithmetic would cause the
 //     8th bit to be 1: x[i] = x[i] + 0x01
 // (3) Only leave the 8th bit by masking with 0x80.
 // (4) If any x[i] is nan, then the whole x != 0.
 
 template<>
-struct CheckBytePack<c10::Float8_e4m3fn, /*EltPerPack*/16> {
-  static __device__ __forceinline__ bool hasNanFP8x8(uint64_t fp8x8) {
+struct HasNanFP8x8<c10::Float8_e4m3fn> {
+  static __device__ __forceinline__ bool check(uint64_t fp8x8) {
     auto t = fp8x8 & 0x7F7F7F7F7F7F7F7FULL;
     auto incremented = t + 0x0101010101010101ULL;
     auto overflow = incremented & 0x8080808080808080ULL;
     return overflow != 0;
   }
+};
+
+// isnan condition for Float8_e5m2:
+// (x & 0x7f) > 0x7c
+// This case does not overflow: 0x7c + 0x03 == 0x7f but adding 0x03 to anything
+// greater than 0x7c will overflow.
+
+template<>
+struct HasNanFP8x8<c10::Float8_e5m2> {
+  static __device__ __forceinline__ bool check(uint64_t fp8x8) {
+    auto t = fp8x8 & 0x7F7F7F7F7F7F7F7FULL;
```

Comment:
Interesting, we can't use std::numerics_limit for either of these FP8 to check NaN equality? Or is this just faster?

Reply:
Ah wait, I see it's packed fp8, nvm.

Reply:
Yeah, speed is one of the reasons -- I don't know how to teach CUDA to apply `isnan()` (or any mapping) to each byte of `fp8x8`, followed by a reduction of the results, and still run fast.

Please let me know if you have any suggestion!

Reply:
@kwen2501 You could use the self comparison value trick to check if NaN, but I'm not sure if that's faster.